addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/stop-ec2-environment/stop-ec2-environment-sc.js

```diff
@@ -0,0 +1,149 @@
+        throw error;
+      }
+    }
+    await this.updateEnvironmentStatus('STOPPING');
+
+    this.state.setKey('STATE_INSTANCE_ID', instanceId);
+
+    return this.wait(5)
+      .maxAttempts(120)
+      .until('checkInstanceStopped')
+      .thenCall('updateEnvironmentStatusToStopped');
+  }
+
+  async checkInstanceStopped() {
+    const instanceId = await this.state.string('STATE_INSTANCE_ID');
+    this.print(`instanceId: [${instanceId}]`);
+
+    const ec2 = await this.getEc2Service();
+    const params = {
+      InstanceIds: [instanceId],
+    };
+
+    let instanceStatusInfo;
+    try {
+      const parmsForDescribe = { ...params, IncludeAllInstances: true };
+      instanceStatusInfo = await ec2.describeInstanceStatus(parmsForDescribe).promise();
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error('describe ec2 instance error: ', error);
+      throw error;
+    }
+
+    const status = _.get(instanceStatusInfo, 'InstanceStatuses[0].InstanceState.Name').toUpperCase();
+
+    if (['SHUTTING-DOWN', 'TERMINATED'].includes(status)) {
+      throw new Error(`EC2 instance [${instanceId}] is in ${status} state and can not be stopped`);
+    }
+
+    return status === 'STOPPED';
+  }
+
+  async getEc2Service() {
+    const [aws] = await this.mustFindServices(['aws']);
+    const [requestContext, RoleArn, ExternalId] = await Promise.all([
+      this.payload.object('requestContext'),
+      this.payload.string('cfnExecutionRole'),
+      this.payload.string('roleExternalId'),
+    ]);
+
+    const sts = new aws.sdk.STS();
+    const {
+      Credentials: { AccessKeyId: accessKeyId, SecretAccessKey: secretAccessKey, SessionToken: sessionToken },
+    } = await sts
+      .assumeRole({
+        RoleArn,
+        RoleSessionName: `RaaS-${requestContext.principalIdentifier.username}`,
+        ExternalId,
+      })
+      .promise();
+
+    return new aws.sdk.EC2({ accessKeyId, secretAccessKey, sessionToken });
+  }
+
+  async updateEnvironmentStatusToStopped() {
+    return this.updateEnvironmentStatus('STOPPED');
+  }
+
+  async updateEnvironmentStatus(status) {
+    const environmentScService = await this.mustFindServices('environmentScService');
+    const id = await this.state.string('STATE_ENVIRONMENT_ID');
+    const requestContext = await this.state.optionalObject('STATE_REQUEST_CONTEXT');
+
+    // SECURITY NOTE
+    // add field to authorize update on behalf of user
+    // this is needed to allow shared envirnments to start/stop by other users
+    requestContext.fromWorkflow = true;
+    const existingEnvRecord = await environmentScService.mustFind(requestContext, { id, fields: ['rev'] });
+
+    const environment = {
+      id,
+      rev: existingEnvRecord.rev || 0,
+      status,
+    };
+    await environmentScService.update(requestContext, environment);
+  }
+
+  async onFail() {
+    return this.updateEnvironmentStatus('STOPPING_FAILED');
```

Comment:
It might be out of scope but can user retry start/stop functionality if they get into this failed state due to network or EC2 issues?

Reply:
Good call! I'll see if I can slip it in this PR :)

Reply:
Looks like I don't have to add this one in. I'll create a separate ticket for it.